[tf.data] graduate ThreadingOptions from experimental to tf.data

kvignesh1420 opened the pull request:
**[UPDATED]** This PR graduates `tf.data.experimental.ThreadingOptions` to `tf.data.ThreadingOptions`.

- [x] deprecate the old endpoint
- [x] update docstring of the class
- [x] Modify the serialization-deserialization implementation of `options.threading` and `options.experimental_threading` for backward-compatibility
- [x] regenerate golden APIs
- [x] add relevant test cases for the new ser-des implementation.
- [x] update RELEASE.md

TEST LOG
```
INFO: Elapsed time: 43.917s, Critical Path: 34.51s
INFO: 351 processes: 130 internal, 221 processwrapper-sandbox.
INFO: Build completed successfully, 351 total actions
//tensorflow/python/data/kernel_tests:options_test                       PASSED in 3.3s

INFO: Build completed successfully, 351 total actions
```

cc: @jsimsa

kvignesh1420 committed:
[tf.data] deprecate experimental ThreadingOptions

kvignesh1420 committed:
modify ThreadingOptions usage in docstring

kvignesh1420 committed:
regenerate golden apis

kvignesh1420 committed:
update RELEASE.md

kvignesh1420 committed:
modify ThreadingOptions target in Options

jsimsa reviewed (changes requested):
We also want to update `tf.data.Options` implementation to add `threading` attribute (in addition to the `experimental_threading` attribute and make sure that the values of the attribute as properly serialized and deserialized. The implementation should probably checked that only one of `threading` and `experimental_threading` has non-default values and use that one when serializing the Python `tf.data.Options` object into a proto.

kvignesh1420 commented:
@jsimsa as per your comment, while converting the `experimental_threading` or `threading` options to proto we can choose based on non-default values. However, how do we decide which one to populate while retrieving from proto? We have to populate both to ensure backward compatibility, but that contradicts the comment.

jsimsa commented:
> @jsimsa as per your comment, while converting the `experimental_threading` or `threading` options to proto we can choose based on non-default values. However, how do we decide which one to populate while retrieving from proto? We have to populate both to ensure backward compatibility, but that contradicts the comment when the users try to convert it to a proto again.

You raise a good point. The short answer is that I am fine with not providing backwards compatibility for getting the value of the threading options (which I expect to be an rare use case as the options are commonly set but user code but rarely read).  So my recommendation would be to only populate the `threading` object and when the `experimental_threading` attribute is access, we should log a warning to inform users that they should be using the `threading` attribute instead.

kvignesh1420 committed:
implement threading option ser-des

kvignesh1420 committed:
set preference order for overwriting threading options

kvignesh1420 committed:
log deprecation warning for experimental_threading

kvignesh1420 commented:
@jsimsa I have taken a slightly different approach in which I set the `experimental_threading` and `threading` values to be the same while ser-des. However, preference is given to `threading` while converting to proto. I have made the modifications to the tests and added a deprecation warning when users set the `experimental_threading` attribute. Let me know how it looks.

kvignesh1420 commented on tensorflow/python/data/util/options.py:
```diff
@@ -60,6 +60,9 @@ def __setattr__(self, name, value):
                        "`tf.data.Dataset.with_options(options)` to set or "
                        "update dataset options.")
     if hasattr(self, name):
+      if name=="experimental_threading":
+        logging.warning("options.experimental_threading is deprecated."
+                        "Use options.threading instead.")
```
An alternative approach would be to:
```
if name=="experimental_threading":
    logging.warning("options.experimental_threading is deprecated."
                    "Use options.threading instead.")
    object.__setattr__(self, "threading", value)
```

and in `dataset_ops.Options` serialize only `threading` to proto. (Although both `threading` and `experimental_threading` will be set and will always be equal during serdes).

@jsimsa WDYT?

jsimsa replied:
I like the alternative. That way we do not need to try to keep `threading` and `experimental_threading` in sync and the `experimental_threading` attribute becomes unused.

jsimsa commented on tensorflow/python/data/experimental/ops/threading_options.py:
```diff
@@ -52,6 +54,14 @@ class ThreadingOptions(options.OptionsBase):
       "The value 0 can be used to indicate that the threadpool size should be "
       "determined at runtime based on the number of available CPU cores.")
 
+  def _has_default_values(self):
```
change the name to `_has_non_default_values()` and the logic accordingly

kvignesh1420 replied:
addressed in the commit.

jsimsa commented on tensorflow/python/data/experimental/ops/threading_options.py:
```diff
@@ -52,6 +54,14 @@ class ThreadingOptions(options.OptionsBase):
+    for attr in attrs:
+      if object.__getattribute__(self, attr) is not None:
+        return False
+    else:
```
the else branch should not be necessary

kvignesh1420 replied:
addressed in the commit

jsimsa commented on tensorflow/python/data/ops/dataset_ops.py:
```diff
@@ -3479,7 +3488,13 @@ def _to_proto(self):
+    # (kvignesh1420): We try to keep the values of `threading` and
+    # `experimental_threading` the same, to prevent unexpected behaviours.
+    if not self.threading._has_default_values():
+      self.experimental_threading = self.threading
```
When you do this, you should check whether `experimental_threading` has any values that are different from `threading` and if so, log a warning with information about which `experimental_threading` options are being ignored.

kvignesh1420 replied:
addressed in the commit

kvignesh1420 replied:
addressed in the commit

jsimsa reviewed (commented):
Thanks Vignesh, did a pass and left some comments.

kvignesh1420 committed:
simplify switch-over to options.threading

kvignesh1420 committed:
ensure backward-compatibility with options.experimental_threading

override experimental_threading opts with threading opts

fix failing ThreadingOption tests

fix failing tests

jsimsa commented on tensorflow/python/data/kernel_tests/options_test.py:
```diff
@@ -201,6 +201,14 @@ def testProtoOptionsDefaultValuesRoundTrip(self):
+  def testThreadingOptionsBackwardCompatibility(self):
+    options = dataset_ops.Options()
+    options.threading.max_intra_op_parallelism = 20
+    self.assertEqual(options.threading, options.experimental_threading)
```
instead, check that both `options.threading.max_intra_op_parallelism` and `options.experimental_threading.max_intra_op_parallelism` is 20

kvignesh1420 replied:
modified the test cases a bit.

kvignesh1420 replied:
```
instead, check that both options.threading.max_intra_op_parallelism and options.experimental_threading.max_intra_op_parallelism is 20
```
This is valid only after ser-des. Adding a new test for that.

jsimsa commented on tensorflow/python/data/kernel_tests/options_test.py:
```diff
@@ -201,6 +201,14 @@ def testProtoOptionsDefaultValuesRoundTrip(self):
+    options.threading.max_intra_op_parallelism = 20
+    self.assertEqual(options.threading, options.experimental_threading)
+    options.experimental_threading.private_threadpool_size = 80
+    self.assertEqual(options.threading, options.experimental_threading)
```
ditto

kvignesh1420 replied:
modified the test cases.

kvignesh1420 commented:
@jsimsa added the logger which warns about the overridden options from `experimental_threading` and modified the test cases a bit. PTAL.

kvignesh1420 committed:
simplify naming convention in threading_options

refactor options.threading serialization

prefer getattr over __getattribute__

regenerate golden apis

jsimsa commented on tensorflow/python/data/experimental/ops/threading_options.py:
```diff
@@ -52,6 +54,15 @@ class ThreadingOptions(options.OptionsBase):
       "The value 0 can be used to indicate that the threadpool size should be "
       "determined at runtime based on the number of available CPU cores.")
 
+  def _get_option_attrs(self):
+    return ["max_intra_op_parallelism", "private_threadpool_size"]
```
You can avoid introducing this method by modifying the `_has_non_default_values` to do the following instead:

```
for attr in filter(lambda x: not x.startswith("_"), dir(self)):
  ...
```

kvignesh1420 replied:
addressed in the commit

jsimsa commented on tensorflow/python/data/ops/dataset_ops.py:
```diff
@@ -3479,7 +3489,24 @@ def _to_proto(self):
+    # (kvignesh1420): We try to keep the values of `threading` and
+    # `experimental_threading` the same, to prevent unexpected behaviours
+    # and ensure backward-compatibility.
+    if self.threading._has_non_default_values():
+      if self.experimental_threading._has_non_default_values():
+        override_attrs = []
+        for attr in self.threading._get_option_attrs():
+          if (object.__getattribute__(self.threading, attr) !=
+              object.__getattribute__(self.experimental_threading, attr)):
+            override_attrs.append(attr)
+        if override_attrs:
+          logging.warning("overriding attr(s) '{}' of experimental_threading "
+                "with respective values in threading.".format(
+                  ",".join(override_attrs)))
```
since we are only checking a small number of values, my suggestion would be to simply check that all attributes of the two objects are identical (instead of checking whether either of them has non-default values)

kvignesh1420 replied:
addressed in the commit

jsimsa commented on tensorflow/python/data/experimental/ops/threading_options.py:
```diff
@@ -52,6 +54,15 @@ class ThreadingOptions(options.OptionsBase):
+
+  def _has_non_default_values(self):
+    for attr in self._get_option_attrs():
+      if object.__getattribute__(self, attr) is not None:
```
use `getattr` instead of `object.__getattribute__`

kvignesh1420 replied:
addressed in the commit

jsimsa commented on tensorflow/python/data/experimental/ops/threading_options.py:
```diff
@@ -52,6 +54,12 @@ class ThreadingOptions(options.OptionsBase):
       "The value 0 can be used to indicate that the threadpool size should be "
       "determined at runtime based on the number of available CPU cores.")
 
+  def _has_non_default_values(self):
+    for attr in filter(lambda opt: not opt.startswith("_"), dir(self)):
+      if getattr(self, attr) is not None:
+        return True
+    return False
+
```
remove as this is not used anywhere anymore

kvignesh1420 replied:
addressed in the commit

jsimsa commented on tensorflow/python/data/ops/dataset_ops.py:
```diff
@@ -3479,7 +3489,20 @@ def _to_proto(self):
+    # (kvignesh1420): We try to keep the values of `threading` and
+    # `experimental_threading` the same, to prevent unexpected behaviours
+    # and ensure backward-compatibility.
+    override_attrs = []
```
instead of collecting a list of overridden and then printing a single warning, log the warning for each attribute, listing the divergent values:

```
for attr in filter(lambda opt: not opt.startswith("_"), dir(self.threading)):
  x, y = getattr(self.threading, attr), getattr(self.experimental_threading, attr)
  if (x != y):
    log.warning("overriding `experimental_threading.{} = {}` with `threading.{} = {}", attr, x, attr, y)
```

kvignesh1420 replied:
addressed in the commit

kvignesh1420 committed:
sanitize options.threading serialization process

rthadur commented:
@kvignesh1420 can you please resolve conflicts ?

kvignesh1420 commented:
@rthadur the conflicts have been resolved.
